### ISSUE TITLE
Transaction _out of gas_ even though it goes through

### ISSUE THREAD

**Issue by fratellopartigiano** [paritytech/parity#8898]
> I'm running:
>
> - **Which Parity version?**: Parity v1.11.1 (tried also 1.10.4)
> - **Which operating system?**: Ubuntu 16.04.4 LTS, kernrel 4.4.0-127-generic
> - **How installed?**: docker, binaries
> - **Are you fully synchronized?**: irrelevant
> - **Which network are you connected to?**: devchain, kovan
> - **Did you try to restart the node?**: irrelevant

> In addition:
> - Solidity 0.4.24 (tried also 0.4.23)

I'm trying to deploy some smart contracts (pasted below) and perform some actions. **Although it runs fine in `kovan` or in `remix`**, it always failes on a development chain (I've tried custom PoW and PoA chainspecs, as well as a copy of kovan chainspec).

The problem appears when executing the `upgradeTo()` function. `web3js` reports the transaction ran out of gas, whereas it in fact goes through (which can be seen by viewing the modified state). **This problem does not appear when I use `web3py` instead.** — therefore this might be `web3js` issue, although given the fact it works on `kovan`, it seems more like being related to `parity`.

**This issue might be related to #8899 .**

*Proxy.sol*
```
pragma solidity ^0.4.24;

// taken from openzeppelin and edited

/**
 * @title Proxy
 * @dev Gives the possibility to delegate any call to a foreign implementation.
 */
contract Proxy {

    address public impl;

    constructor() public {
    }

    function upgradeTo(address newimpl) external onlyOwner {
	impl = newimpl;
    }
    
    function implementation() public view returns (address) {
	return impl;
    }

    function () payable public {
	address _impl = implementation();
	require(_impl != address(0));

	assembly {
	    let ptr := mload(0x40)
	    calldatacopy(ptr, 0, calldatasize)
	    let result := delegatecall(gas, _impl, ptr, calldatasize, 0, 0)
	    let size := returndatasize
	    returndatacopy(ptr, 0, size)

	    switch result
	    case 0 { revert(ptr, size) }
	    default { return(ptr, size) }
	}
    }
}
```

*LogicPrimary.sol*
```
pragma solidity ^0.4.24;

contract LogicPrimary {

    address public impl;
    
    constructor() public {
    }

    function doSomething() public pure returns(uint32) {
	return 1;
    }
}
```

*Testing script*
```
#!/bin/node

const Web3 = require('web3');
var web3 = new Web3(Web3.givenProvider || "ws://localhost:8548");

let logicContractInstance = undefined;
let proxyContractInstance = undefined;

let proxyData = "0x608060405234801561001057600080fd5b50336000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506107db806100606000396000f3006080604052600436106100c5576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680630dbe671f146101345780633659cfe61461016b5780634df7e3d0146101ae5780634f103353146101e55780635c60da1b146102185780638abf60771461026f5780638da5cb5b146102c6578063a1c519151461031d578063a2375d1e14610354578063bd8aad271461038b578063c3da42b8146103be578063d46300fd146103f5578063dca75ea31461042c575b60006100cf61046f565b9050600073ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff161415151561010d57600080fd5b60405136600082376000803683855af43d806000843e8160008114610130578184f35b8184fd5b34801561014057600080fd5b50610149610499565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561017757600080fd5b506101ac600480360381019080803573ffffffffffffffffffffffffffffffffffffffff1690602001909291905050506104af565b005b3480156101ba57600080fd5b506101c361054e565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b3480156101f157600080fd5b50610216600480360381019080803563ffffffff169060200190929190505050610564565b005b34801561022457600080fd5b5061022d61046f565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561027b57600080fd5b506102846105e3565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b3480156102d257600080fd5b506102db610609565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561032957600080fd5b5061033261062e565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561036057600080fd5b50610369610647565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561039757600080fd5b506103bc600480360381019080803563ffffffff169060200190929190505050610660565b005b3480156103ca57600080fd5b506103d36106df565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561040157600080fd5b5061040a6106f5565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561043857600080fd5b5061046d600480360381019080803563ffffffff169060200190929190803563ffffffff16906020019092919050505061070e565b005b6000600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b600060149054906101000a900463ffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614151561050a57600080fd5b80600160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b600060189054906101000a900463ffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161415156105bf57600080fd5b80600060186101000a81548163ffffffff021916908363ffffffff16021790555050565b600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60008060189054906101000a900463ffffffff16905090565b600080601c9054906101000a900463ffffffff16905090565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161415156106bb57600080fd5b80600060146101000a81548163ffffffff021916908363ffffffff16021790555050565b6000601c9054906101000a900463ffffffff1681565b60008060149054906101000a900463ffffffff16905090565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614151561076957600080fd5b81600060146101000a81548163ffffffff021916908363ffffffff16021790555080600060186101000a81548163ffffffff021916908363ffffffff16021790555050505600a165627a7a723058203429f6be282f83b24d3ef9af2b5ce70595252f9fde25bfb7119763664ce0ec5a0029";
let proxyAbi = JSON.parse("[{\"constant\":true,\"inputs\":[],\"name\":\"a\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"newimpl\",\"type\":\"address\"}],\"name\":\"upgradeTo\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"b\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_b\",\"type\":\"uint32\"}],\"name\":\"setB\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"implementation\",\"outputs\":[{\"name\":\"\",\"type\":\"address\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"impl\",\"outputs\":[{\"name\":\"\",\"type\":\"address\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"owner\",\"outputs\":[{\"name\":\"\",\"type\":\"address\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getB\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getC\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_a\",\"type\":\"uint32\"}],\"name\":\"setA\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"c\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getA\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_a\",\"type\":\"uint32\"},{\"name\":\"_b\",\"type\":\"uint32\"}],\"name\":\"init\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"constructor\"},{\"payable\":true,\"stateMutability\":\"payable\",\"type\":\"fallback\"}]");

let logicData = "0x608060405234801561001057600080fd5b50336000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550610643806100606000396000f3006080604052600436106100ba576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680630dbe671f146100bf5780634df7e3d0146100f65780634f1033531461012d57806382692679146101605780638abf6077146101975780638da5cb5b146101ee578063a1c5191514610245578063a2375d1e1461027c578063bd8aad27146102b3578063c3da42b8146102e6578063d46300fd1461031d578063dca75ea314610354575b600080fd5b3480156100cb57600080fd5b506100d4610397565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561010257600080fd5b5061010b6103ad565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561013957600080fd5b5061015e600480360381019080803563ffffffff1690602001909291905050506103c3565b005b34801561016c57600080fd5b50610175610442565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b3480156101a357600080fd5b506101ac61044b565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b3480156101fa57600080fd5b50610203610471565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561025157600080fd5b5061025a610496565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561028857600080fd5b506102916104af565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b3480156102bf57600080fd5b506102e4600480360381019080803563ffffffff1690602001909291905050506104c8565b005b3480156102f257600080fd5b506102fb610547565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561032957600080fd5b5061033261055d565b604051808263ffffffff1663ffffffff16815260200191505060405180910390f35b34801561036057600080fd5b50610395600480360381019080803563ffffffff169060200190929190803563ffffffff169060200190929190505050610576565b005b600060149054906101000a900463ffffffff1681565b600060189054906101000a900463ffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614151561041e57600080fd5b80600060186101000a81548163ffffffff021916908363ffffffff16021790555050565b60006001905090565b600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60008060189054906101000a900463ffffffff16905090565b600080601c9054906101000a900463ffffffff16905090565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614151561052357600080fd5b80600060146101000a81548163ffffffff021916908363ffffffff16021790555050565b6000601c9054906101000a900463ffffffff1681565b60008060149054906101000a900463ffffffff16905090565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161415156105d157600080fd5b81600060146101000a81548163ffffffff021916908363ffffffff16021790555080600060186101000a81548163ffffffff021916908363ffffffff16021790555050505600a165627a7a7230582067deb3b31d9e361ba68702961985f573d9a6a6987c75de36fbcc0248ea1732890029";
let logicAbi = JSON.parse("[{\"constant\":true,\"inputs\":[],\"name\":\"a\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"b\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_b\",\"type\":\"uint32\"}],\"name\":\"setB\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"doSomething\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"pure\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"impl\",\"outputs\":[{\"name\":\"\",\"type\":\"address\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"owner\",\"outputs\":[{\"name\":\"\",\"type\":\"address\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getB\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getC\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_a\",\"type\":\"uint32\"}],\"name\":\"setA\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"c\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"getA\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_a\",\"type\":\"uint32\"},{\"name\":\"_b\",\"type\":\"uint32\"}],\"name\":\"init\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"constructor\"}]");

console.log("// Deploy logic contract");
const logicContract = new web3.eth.Contract(logicAbi);
const fromAddress = ""
logicContract.deploy({ data: logicData }).send({from:fromAddress})
    .then(logicInstance => {
	console.log("// deploy proxy instance");
	logicContractInstance = logicInstance;
	const proxyContract = new web3.eth.Contract(proxyAbi);
	return proxyContract.deploy({ data: proxyData }).send({from:fromAddress});
    }).then(proxyInstance => {
	console.log("// Set proxy upgardeTo to point to address of logic contract");
	console.log(proxyInstance.options.address);
	proxyContractInstance = proxyInstance;
	console.log(logicContractInstance.options.address);
	return proxyInstance.methods.upgradeTo(logicContractInstance.options.address).send({from:fromAddress});
    }).then(receipt => {
	console.log("// call the proxy contract as a logic contract instance")
	let proxyInstanceAsLogic = new web3.eth.Contract(logicAbi, proxyContractInstance.options.address);
	return proxyInstanceAsLogic.methods.doSomething().call();
    }).then(returnValue => {
	console.log(returnValue);
    }).catch(err => console.error("error"+err));
```

*Example output*
```
// Deploy logic contract
// deploy proxy instance
// Set proxy upgardeTo to point to address of logic contract
0x8F8DEC1450d930f1E2A0f04074c1a871D74f62F3
0x1fA7e7d810DD8A52dF276328fBC5eA7Fd374254A
errorError: Transaction ran out of gas. Please provide more gas:
{
  "blockHash": "0x54ac608c6f38254040d216d21e3709656c0447fb97da1ca87648d765d0c1d421",
  "blockNumber": 5473,
  "contractAddress": null,
  "cumulativeGasUsed": 43442,
  "gasUsed": 43442,
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "root": "0x8ebf8bccb2eac8c53d0873ae22d5aa62b3e3771f75d347968990f2a3fd4a1d77",
  "status": false,
  "transactionHash": "0x9dbb3ad21b9ea54e87fe5d881393738a9f350172c0570b28ca359e5839c4ca03",
  "transactionIndex": 0,
  "events": {}
}
```

**Comment by Tbaut** [comment]
The fact that you do not specify any gas when sending transaction leaves parity set it for you. Depending on the network you will be on, the gas and gasPrice will surely differ. web3 and web3py also probably have different behavior regarding this. Could you track what values are used and see if this could explain to the different behaviors?

**Comment by Tbaut** [comment]
Also please upgrade to the security fix release *1.11.3*: https://github.com/paritytech/parity/releases

**Comment by fratellopartigiano** [comment]
It always says `"gasUsed": 43442`, I tried also specifying the gas manually to higher values, but with no effect. The 43442 is the actual cost of the transaction, as I get this value when I execute on kovan, as well as when I execute with web3py on any chain.

I'm afraid 1.11.3 is not available dockerized yet, as I get:
```
docker pull parity/parity:latest
latest: Pulling from parity/parity
Digest: sha256:d298c50a32a6e4aa19afd94020be4a1962c7c270c934e11e97b8af6979cfa081
Status: Image is up to date for parity/parity:latest
```
and this is v1.11.1-beta-6654d02-20180515.

**Comment by fratellopartigiano** [comment]
Solved along with #8899.

**Comment by Tbaut** [comment]
@fratellopartigiano is dockerised but we currently have a problem with our docker hub (will be solved in the next release). Please use the tags `parity/parity:v1.11.3` in the mean time
